4/7 blog (#1167)
Azure/azure-sdk: _posts/2020-04-07-java-eventprocessorclient.md

---
title: Process data in real-time with Java and Azure Event Hubs
layout: post
date: 07 April 2020
sidebar: releases_sidebar
author_github: conniey
repository: azure/azure-sdk
---

In my position with the Azure SDK team, I'm fortunate to talk to a lot of customers about their needs.  Most applications can be categorized into big buckets where we can see the same design patterns repeated.  One of those categories is data processing.  Let's take an example.  A farm needs to understand the micro-climate weather patterns across the land under management.  This allows them to predict where they should water their crops more, and where the best grazing for their animals are, among other things.  To do that, they set up small weather stations at regular intervals.  The weather stations send their data to "the cloud" for processing and display.

How might this work?

![]({{ site.base_url }}/images/posts/04072020-image1.png)

> Each base station has a number of feeds from the instruments, each of which could provide information at different rates.  This is all fed into an [Azure Event Hub](https://azure.microsoft.com/en-us/services/event-hubs/).  Azure Event Hubs are a scalable messaging system primarily used for data ingestion, allowing you to analyze millions of data points.  You could use the same system for analyzing marketing data coming from a mobile application, shopping data coming from the scanners at a grocery store, or security devices monitoring millions of homes.  We've learned a few things about this, and have come up with some best practices to follow.  Although you can use any of our languages for this (.NET, Python, Java, or TypeScript), we will be using Java for the examples today.

Reviewer: Please use JavaScript/TypeScript, not just TypeScript.  Thanks

> Each base station has a number of feeds from the instruments, each of which could provide information at different rates.  This is all fed into an [Azure Event Hub](https://azure.microsoft.com/en-us/services/event-hubs/).  Azure Event Hubs are a scalable messaging system primarily used for data ingestion, allowing you to analyze millions of data points.  You could use the same system for analyzing marketing data coming from a mobile application, shopping data coming from the scanners at a grocery store, or security devices monitoring millions of homes.  We've learned a few things about this, and have come up with some best practices to follow.  Although you can use any of our languages for this (.NET, Python, Java, or TypeScript), we will be using Java for the examples today.

Reviewer: Consider linking to the SDK repos for each lang mentioned

## Send data in batches most of the time

In a data ingestion pipeline, individual components that are sending data often have more resources available than the central receiver has, so you want to optimize the pipeline to ensure the central receiver is doing less work.  This means sending data in batches.  Data is sent to an Event Hub using [AMQP](https://en.wikipedia.org/wiki/Advanced_Message_Queuing_Protocol), which provides for batching in the protocol.

Applications using Event Hubs are split into producers (the app that sends data) and consumers (the app that receives data).  Take a look at this example:

```java
public class TemperatureSender {

>   private EventHubProducerClient sender;
>   private final ObjectMapper serializer = new ObjectMapper;
>   private TemperatureDevice device;

Reviewer 2: ```suggestion
  private final EventHubProducerClient sender;
  private final ObjectMapper serializer = new ObjectMapper();
  private final TemperatureDevice device;
```

  public TemperatureSender(String namespace, String hubName, TemperatureDevice device) {
    TokenCredential credential = new DefaultAzureCredentialBuilder().build();
    this.sender = new EventHubClientBuilder()
      .credential(namespace, hubName, credential)
      .buildProducerClient();
    this.device = device;
  }

  // Rest of the class
}
```

This is a fictitious class that will send temperature data from a connected device to the Event Hub.  You specify the Event Hub namespace and name (both obtained when provisioning the Event Hubs) to configure the connection.

> 

Reviewer: I would prefer to show the code instead of not shown.  I'd also accompany this with a sample so they can download and run.

The `TemperatureDevice` class (not shown) is responsible for reading data from the device.  How do we send data?  Let's create a method that reads a temperature and stores it in a batch.
When the batch is full, it will send the data to the Event Hub.

```java

>    private EventDataBatch currentBatch = null;

Reviewer 3: decrease indentation for these lines by 4 spaces

    public sendTemperatureData() {
      // Get the temperature from the device
      Temperature measure = this.device.getTemperature();
      // Serialize the temperature as a string value

>       String json = this.serializer.writeValueAsString(measure.toString());

Reviewer 2: Let's make the assumption that the `Temperature` class is probably annotated to be JSON serialized.

```suggestion
      String json = this.serializer.writeValueAsString(measure);
```

In our example above, the sender selects the partition to send an event batch to in a round-robin basis, thus spreading the load between all partitions.  You can select a specific partition by using `.setPartitionId()` on the `CreateBatchOptions`.  Partition IDs are strings and set by the service.  Get the list of partition Ids using `this.sender.getPartitionIds()`.

## Use the EventProcessorClient

Now that the data we want to process is in the Event Hub, it's time to look at processing it.  Let's think about what you have to think about when processing events:

1. The event processor (like all cloud-native services) can be restarted at any point, so you must build in resiliency.
2. You don't want to repeatedly process the same events by starting from the beginning of the stream each time, so you must record a checkpoint every now and then.
3. Even though data is sent in batches, the batch sizes may vary.  Your processing strategy may be different from your sending strategy.

> There is a low-level client for processing events.  Unless you are very concerned about latency (to the point where you really want to control when and how many events you pull off the Event Hub and skip events if you are getting behind), you will want to use the `EventProcessorClient`.

Reviewer 2: Could we flip this, I think it is better to state what we are going to talk about first rather than off hand mentioning this other client and never saying anything else about it.

```suggestion
Generally, for processing events you will want to use the `EventProcessorClient`, for scenarios where you are very concerned about latency there is a low-level client. You will want to investigate using the low-level client where you are at the point where you really want to control when and how many events you pull off the Event Hub and skipping events if you are getting behind in processing.
```

> The EventProcessorClient is a higher-level construct wrapping the low-level client.  It does all the "boiler-plate" event management for you; for example, coordinating processing between multiple hosts, allowing other processor clients to "take over" if you need to move the processing to another host (for example, to support zero-down-time upgrades), and storing checkpoint information.

Reviewer 2: ```suggestion
The `EventProcessorClient` is a higher-level construct wrapping the low-level client.  It does all the "boiler-plate" event management for you; for example, coordinating processing between multiple hosts, allowing other processor clients to "take over" if you need to move the processing to another host (for example, to support zero-down-time upgrades), and storing checkpoint information.
```
